CONTRIBUTING.md

```diff
@@ -0,0 +1,136 @@
+# Contribution guide
+
+The codebase is maintained using the "contributor workflow" where anyone can
+contribute patch proposals using
+[pull requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests).
+
+To contribute a patch, the workflow is as follows:
+
+1. Fork the repository.
+2. Create a topic branch.
+3. Commit patches to the branch.
+4. Push changes to the fork.
+5. Create a pull request to merge the branch of the fork into this repository.
+6. If you had someone specifically in mind, ask them to review the pull request.
+Otherwise, just wait for a code review: most members with merge permissions
+receive notifications for newly created pull requests.
+7. Address review comments, if any.
+8. Merge and submit the pull request. If you don't have merge permissions,
+a reviewer will do it for you.
+
+> **NOTE:** Before starting any coding work, please make sure to discuss the
+> issue first to agree on the optimal way of solving a particular problem.
+> If no issue exists, create one unless the change is trivial, such as
+> correcting typos and code code formatting.
+>
+> If you don't know where to start, look for issues labeled
+> with `good-first-issue`.
+
+## Branches
+
+When creating a new branch, give it a meaningful name. For example, `new-coin-xyz`.
+Later, when a pull request is finally merged in a non fast-forward way, the merge
+commit would read:
+
+    Merge branch new-coin-xyz
+
+This makes it easier to read the commit history.
+
+## Git commits
+
+In general,
+[commits should be atomic](https://en.wikipedia.org/wiki/Atomic_commit#Atomic_commit_convention)
+and diffs should be easy to read. Please, do not mix style, formatting or moving
+code around with the actual code changes in the same commit. If a change is
+larger than 500 lines, consider creating multiple pull requests or multiple
+commits depending on how closely the patches are related to each other.
+
+For example, if a change is much easier to do after a certain code reformatting,
+make the reformatting in a first commit and the actual change in the second.
+On the other hand, if a change includes multiple backend commits to create
+a new API endpoint as well as some UI modifications on the frontend, consider
+splitting backend and frontend into separate pull requests.
+
+A commit message consists of a subject, the first line, and a body separated
```

Comment:
maybe simpler to link to one of the many commit messages guides, e.g. the one from Go? can add anything on top below.

Reply:
i'm a bit hesitant to link to go or blog posts - those links can move away and i wouldn't want this guide to point to broken links. i think the guide covers enough of the basics.

i figured we can always update this if it turns out to not be enough. let's see.